File: DataStructure/src/HashTable/HashTable.java
package HashTable;

public class HashTable implements HashTableInterface {
    private static final DataItem DELETED = new DataItem(-1);
    private DataItem[] hashArray;

    public HashTable(int initialCapacity) {
        hashArray = new DataItem[initialCapacity];
    }

    private int hashFunc(int key) {
        return key % hashArray.length;
    }

    @Override
    public boolean search(int key) {
        int hashVal = hashFunc(key);
        while (hashArray[hashVal] != null) {
            if (hashArray[hashVal].key == key) {
                return true;
            }
            hashVal ++;
            hashVal %= hashArray.length;
        }
        return false;
    }

    @Override
    public int delete(int key) {
        int hashVal = hashFunc(key);
        while (hashArray[hashVal] != null) {
            if (hashArray[hashVal].key == key) {
                // should return value from data structure
                int val = hashArray[hashVal].key;
                hashArray[hashVal] = DELETED;
                return val;
            }
            hashVal ++;
            hashVal %= hashArray.length;
        }
        return -1;
    }

    @Override
    public void insert(int key) {
        DataItem item = new DataItem(key);
        int hashVal = hashFunc(key);

        while (hashArray[hashVal] != null && hashArray[hashVal] != DELETED) {
            hashVal ++;
            hashVal %= hashArray.length;
        }
        hashArray[hashVal] = item;
    }

    private static class DataItem {
        private int key;

        DataItem(int key) {
            this.key = key;
        }
    }
}

File: DataStructureGo/hashtable.go
package main

type DataItem struct {
	key int
}

type HashTable struct {
	hashArray []*DataItem
	deleted   *DataItem
}

func MakeHashTable(capacity int) HashTable {
	return HashTable{make([]*DataItem, capacity), &DataItem{-1}}
}

func (table *HashTable) hash(key int) int {
	return key % len(table.hashArray)
}

func (table *HashTable) search(key int) bool {
	hashVal := table.hash(key)
	for table.hashArray[hashVal] != nil {
		if table.hashArray[hashVal].key == key {
			return true
		}
		hashVal++
		hashVal %= len(table.hashArray)
	}
	return false
}

func (table *HashTable) insert(key int) {
	item := DataItem{key}
	hashVal := table.hash(key)
	for table.hashArray[hashVal] != nil && table.hashArray[hashVal] != table.deleted {
		hashVal++
		hashVal %= len(table.hashArray)
	}
	table.hashArray[hashVal] = &item
}

func (table *HashTable) delete(key int) int {
	hashVal := table.hash(key)
	for table.hashArray[hashVal] != nil {
		if table.hashArray[hashVal].key == key {
			val := table.hashArray[hashVal].key
			table.hashArray[hashVal] = table.deleted
			return val
		}
		hashVal++
		hashVal %= len(table.hashArray)
	}
	return table.deleted.key
}
